ntoskrnl/io/iomgr/deviface.c

```diff
@@ -304,7 +305,191 @@ IoGetDeviceInterfaceAlias(IN PUNICODE_STRING SymbolicLinkName,
+    UNICODE_STRING GuidString, AliasGuidString;
+    UNICODE_STRING DeviceString, ReferenceString;
+    PKEY_VALUE_FULL_INFORMATION kvInfo;
+    HANDLE InstanceHandle;
+    USHORT i, ReferenceStringOffset;
+    NTSTATUS Status;
+
+    SymbolicLink.Buffer = LinkNameNoPrefix.Buffer = DeviceString.Buffer = ReferenceString.Buffer = GuidString.Buffer = AliasGuidString.Buffer = NULL;
+
+    DPRINT1("IoGetDeviceInterfaceAlias() : symbolic link '%wZ'\n", SymbolicLinkName);
+
+    /* Sanity check */
+    if (SymbolicLinkName == NULL || AliasInterfaceClassGuid == NULL)
+    {
+        DPRINT1("IoGetDeviceInterfaceAlias() : invalid symbolic link or alias class GUID\n");
+        return STATUS_INVALID_HANDLE;
+    }
+
+    /* Symbolic link name looks like \??\ACPI#PNP0501#1#{GUID}\ReferenceString */
+    /* Make sure it starts with the expected prefix */
+    if (!RtlPrefixUnicodeString(&DosDevicesPrefix1, SymbolicLinkName, FALSE) &&
+        !RtlPrefixUnicodeString(&DosDevicesPrefix2, SymbolicLinkName, FALSE))
+    {
+        DPRINT1("IoSetDeviceInterfaceState() invalid link name '%wZ'\n", SymbolicLinkName);
+        return STATUS_INVALID_PARAMETER;
+    }
+
+    /* Make a version without the prefix for further processing */
+    ASSERT(DosDevicesPrefix1.Length == DosDevicesPrefix2.Length);
+    ASSERT(SymbolicLinkName->Length >= DosDevicesPrefix1.Length);
+    LinkNameNoPrefix.Buffer = SymbolicLinkName->Buffer + DosDevicesPrefix1.Length / sizeof(WCHAR);
+    LinkNameNoPrefix.Length = SymbolicLinkName->Length - DosDevicesPrefix1.Length;
+    LinkNameNoPrefix.MaximumLength = LinkNameNoPrefix.Length;
+
+    /* Find the reference string, if any */
+    for (i = 0; i < LinkNameNoPrefix.Length / sizeof(WCHAR); i++)
+    {
+        if (LinkNameNoPrefix.Buffer[i] == L'\\')
+        {
+            break;
+        }
+    }
+    ReferenceStringOffset = i * sizeof(WCHAR);
+
+    /* The GUID is before the reference string or at the end */
+    ASSERT(LinkNameNoPrefix.Length >= ReferenceStringOffset);
+    if (ReferenceStringOffset < GUID_STRING_BYTES + sizeof(WCHAR))
+    {
+        DPRINT1("IoSetDeviceInterfaceState() invalid link name '%wZ'\n", SymbolicLinkName);
+        return STATUS_INVALID_PARAMETER;
+    }
+
+    GuidString.Buffer = LinkNameNoPrefix.Buffer + (ReferenceStringOffset - GUID_STRING_BYTES) / sizeof(WCHAR);
+    GuidString.Length = GUID_STRING_BYTES;
+    GuidString.MaximumLength = GuidString.Length;
+
+    /* Get the device and reference strings from symbolic link by the same way */
+    // TODO: Add some more sanity checks
+    ReferenceString.Buffer = LinkNameNoPrefix.Buffer + ReferenceStringOffset / sizeof(WCHAR);
+    ReferenceString.Length = LinkNameNoPrefix.Length - ReferenceStringOffset;
+    ReferenceString.MaximumLength = ReferenceString.Length;
+
+    // FIXME: Calculation seems to be wrong here
+    DeviceString.Buffer = LinkNameNoPrefix.Buffer + (GuidString.Length - ReferenceString.Length) / sizeof(WCHAR);
```

Comment:
I doubt: how I can calculate the buffer properly here?
The function receives symolic link like this:
`\??\PCI#VEN_8086&DEV_2415&SUBSYS_00008086&REV_01#3&609b8881&05#{6994ad04-93ef-11d0-a3cc-00a0c9223196}\Wave`
but device string from it comes out as:
`8086&REV_01#3&609b8881&05#{6994ad04-93ef-11d0-a3cc-00a0c922`
which is a part of device string + old GUID string without `}` bracket.
What I'm doing wrong here? :thinking:

Reply:
I'm also wondering what happens when the symbolic string has different form (that's also a question for @Extravert-ir or @binarymaster, if you know): If the string starts with `PCI#` but followed with something else, or not with `PCI#` at all: What would happen for the resulting device string?

Reply:
Fixed this.